src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+
+                var chosenItems = FilterMethod(matchingItems, _filterText);
+
+                int selectedItemIndex;
+                VSCompletionItem? uniqueItem = null;
+                MatchResult<VSCompletionItem> bestOrFirstMatchResult;
+
+                if (chosenItems.Length == 0)
+                {
+                    // We do not have matches: pick the one with longest common prefix or the first item from the list.
+                    selectedItemIndex = 0;
+                    bestOrFirstMatchResult = items[0];
+
+                    var longestCommonPrefixLength = bestOrFirstMatchResult.RoslynCompletionItem.FilterText.GetCaseInsensitivePrefixLength(_filterText);
+
+                    for (var i = 1; i < items.Count; ++i)
+                    {
+                        var item = items[i];
+                        var commonPrefixLength = item.RoslynCompletionItem.FilterText.GetCaseInsensitivePrefixLength(_filterText);
+
+                        if (commonPrefixLength > longestCommonPrefixLength)
+                        {
+                            selectedItemIndex = i;
+                            bestOrFirstMatchResult = item;
+                            longestCommonPrefixLength = commonPrefixLength;
+                        }
+                    }
+                }
+                else
+                {
+                    var recentItems = _recentItemsManager.RecentItems;
+
+                    // Of the items the service returned, pick the one most recently committed
+                    var bestItem = GetBestCompletionItemBasedOnMRU(chosenItems, recentItems);
+
+                    // Determine if we should consider this item 'unique' or not.  A unique item
+                    // will be automatically committed if the user hits the 'invoke completion' 
+                    // without bringing up the completion list.  An item is unique if it was the
+                    // only item to match the text typed so far, and there was at least some text
+                    // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+                    // like "WriteLine" since no filter text has actually been provided.  However,
+                    // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+                    for (selectedItemIndex = 0; selectedItemIndex < items.Count; ++selectedItemIndex)
+                    {
+                        if (Equals(items[selectedItemIndex].RoslynCompletionItem, bestItem))
+                            break;
+                    }
+
+                    Debug.Assert(selectedItemIndex < items.Count);
+
+                    bestOrFirstMatchResult = items[selectedItemIndex];
+                    var deduplicatedListCount = matchingItems.Count(r => !r.RoslynCompletionItem.IsPreferredItem());
+                    if (deduplicatedListCount == 1 &&
+                        _filterText.Length > 0)
```

Comment:
we should do the _filterTExt check first.  this not only matches what the commetn says above, but it also avoids the wasted .Count work for when nothing has been typed.